WcaOnRails/lib/competition_results_validator.rb

```diff
@@ -0,0 +1,606 @@
+            end
+          else
+            # Handle both cumulative for a single round or multiple round by doing the following:
+            #  - gather all solve times for all the rounds (necessitate to map round's WCIF id to "our" round ids)
+            #  - check the sum is below the limit
+            #  - check for any suspicious DNF result
+
+            # Match wcif round ids to "our" ids
+            cumulative_round_ids = cumulative_wcif_round_ids.map do |wcif_id|
+              parsed_wcif_id = Round.parse_wcif_id(wcif_id)
+              # Get the actual round_id from our expected rounds by id
+              @expected_rounds_by_ids.select do |id, round|
+                round.event.id == parsed_wcif_id[:event_id] && round.number == parsed_wcif_id[:round_number]
+              end.first[0]
+            end
+
+            # Get all solve times for all cumulative rounds for the current person
+            all_results_for_cumulative_rounds = cumulative_round_ids.map do |id|
+              results_by_round_id[id].find { |r| r.personId == result.personId }
+            end.compact.map(&:solve_times).flatten
+            completed_solves_for_rounds = all_results_for_cumulative_rounds.select(&:complete?)
+            number_of_dnf_solves = all_results_for_cumulative_rounds.select(&:dnf?).size
+            sum_of_times_for_rounds = completed_solves_for_rounds.map(&:time_centiseconds).reduce(&:+) || 0
+
+            # Check the sum is below the limit
+            if sum_of_times_for_rounds > time_limit_for_round.centiseconds
+              @errors[:results] << format(RESULTS_OVER_CUMULATIVE_TIME_LIMIT_ERROR, round_ids: cumulative_round_ids.join(","), person_name: person_info.name, time_limit: time_limit_for_round.to_s(round_info))
+            end
+
+            # Check for any suspicious DNF
+            # Compute avg time per solve for the competitor
+            avg_per_solve = sum_of_times_for_rounds.to_f / completed_solves_for_rounds.size
+            # We want to issue a warning if the estimated time for all solves + DNFs goes roughly over the cumulative time limit by at least 10% (to reduce false positive).
+            if (number_of_dnf_solves + completed_solves_for_rounds.size) * avg_per_solve >= 1.1 * time_limit_for_round.centiseconds
+              @warnings[:results] << format(SUSPICIOUS_DNF_WARNING, round_ids: cumulative_round_ids.join(","), person_name: person_info.name)
+            end
+          end
+        end
+
+        if result.eventId == "333mbf"
+          completed_solves.each do |solve_time|
+            time_limit_seconds = [3600, solve_time.attempted * 600].min
+            if solve_time.time_seconds > time_limit_seconds
+              @warnings[:results] << format(MBF_RESULT_OVER_TIME_LIMIT_WARNING, round_id: round_id, result: solve_time.clock_format, person_name: person_info.name)
+            end
+          end
+        end
```

Comment:
Isn't this somehow validated by `validate :multiblind_time_limit` in `SolveTime`?

Reply:
No, this is actually a bit tricky: although the time limit for MBF is 10 minutes per cube up to 60 minutes, it is possible to have (valid) time over this time limit after applying penalties (such as misalignment on cubes).
It's basically impossible currently to tell appart an invalid time from a valid time over the time limit. We don't want to prevent the Delegates from importing solvetime with time over the time limit (which is why we arbitrarily add 30s to the time limit [here](https://github.com/thewca/worldcubeassociation.org/pull/3127/files#diff-d057259b40060c1a85b0c41863c96fbfR255)), but we do want to display a warning to make sure the Delegate checks that this is indeed due to penalties.